### PR TITLE
new flag: `-strict`

### PR DESCRIPTION
**Pull request by AkihiroSuda**
This flag treats non-trivial warning as an error.

As discussed on Slack, this would be useful for detecting invalid vendoring on CI.

cc @moby/moby-maintainers

TODO: add more warnings (in separate PRs)

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

### CODE REVIEW

File: warn.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
+package main
```
Comment (LK4D4) [inline warn.go]: I think it's too much for only one occurrence of `WarnTrivial`. Can we just write some structrure like:
```
type warningCollector struct {
    check bool
    warnings []string
}
```
And use just printf for trivial case.

**Comment by LK4D4** [comment]
@AkihiroSuda I think flag name maybe not right. I'd expect `-check` to do just check, this is more like `-strict`. WDYT?

**Comment by AkihiroSuda** [comment]
updated PR

File: main.go [hunk 2]
```diff
@@ -117,7 +119,7 @@ func validateDeps(deps []depEntry) error {
 		if len(rootDeps) == 1 {
 			d := rootDeps[0]
 			if d.importPath != r {
-				log.Printf("WARNING: package %s is not root import, should be %s", d.importPath, r)
```
Comment (LK4D4) [inline main.go]: Warnings in `validateDeps` are fatal already.

File: main.go [hunk 3]
```diff
@@ -80,7 +82,7 @@ func validateArgs() {
 func checkUnused(deps []depEntry, vd string) {
 	for _, d := range deps {
 		if _, err := os.Stat(filepath.Join(vd, d.importPath)); err != nil && os.IsNotExist(err) {
-			log.Printf("WARNING: package %s is unused, consider removing it from vendor.conf", d.importPath)
+			Warnf("package %s is unused, consider removing it from vendor.conf", d.importPath)
```
Comment (LK4D4) [inline main.go]: So, if I understand correctly the whole PR is about this message, right?

Comment (AkihiroSuda) [inline main.go]: And the `dependency is not vendored` warning in pkg.go

File: main.go (hunk 2 above)
Comment (AkihiroSuda) [inline main.go]: Yes, but I think it is good to use `Warnf` consistently across all the non-trivial warnings

**Comment by LK4D4** [comment]
Okay, makes sense.
Can you pls add test and docs in readme?

**Comment by AkihiroSuda** [comment]
done

File: README.md [hunk 4]
```diff
@@ -12,6 +12,7 @@ exactly what they say on the tin.
   for running tests in a vendored project or otherwise ensuring that some
   important files are retained after `vndr` is done cleaning unused files from
   your `vendor/` directory.
+* `-strict` exists with non-zero status on non-trivial warning
```
Comment (LK4D4) [inline README.md]: typo `exits`

File: test/vndr_test.go [hunk 5]
```diff
@@ -341,4 +342,22 @@ func TestUnused(t *testing.T) {
...
+
+func getExitCode(t *testing.T, err error) int {
+	if err == nil {
+		return 0
+	}
+	exitError, ok := err.(*exec.ExitError)
+	if !ok {
+		t.Fatal("expected *os.ExitError")
+	}
+	status, ok := exitError.Sys().(syscall.WaitStatus)
```
Comment (LK4D4) [inline test/vndr_test.go]: ExitError has `Success()` method

**Pull request merged**